lowball_ad_auth_provider/auth_provider.py

```diff
@@ -0,0 +1,268 @@
+        self._service_account_password = value
+
+    @property
+    def ignore_ssl_cert_errors(self):
+        return self._ignore_ssl_cert_errors
+
+    @ignore_ssl_cert_errors.setter
+    def ignore_ssl_cert_errors(self, value):
+        if not isinstance(value, bool):
+            raise ValueError("ignore_ssl_cert_errors must be a boolean")
+        self._ignore_ssl_cert_errors = value
+
+    @property
+    def role_mappings(self):
+        return self._role_mappings
+
+    @role_mappings.setter
+    def role_mappings(self, value):
+        if value is None:
+            value = {}
+        if not isinstance(value, dict):
+            raise ValueError("Invalid role mappings. Must be a dictionary of {str: [str, str],..}")
+        for key, mappings in value.items():
+            if not isinstance(key, str) or not isinstance(mappings, list) or not all(isinstance(group, str) for group in mappings):
+                raise ValueError("Invalid role mappings. Must be a dictionary of {str: [str, str],..}")
+
+        self._role_mappings = value
+
+    def get_server(self):
+
+        if self.use_ssl:
+            if self.ignore_ssl_cert_errors:
+                tls_configuration = Tls(validate=ssl.CERT_NONE)
+                return Server(host=self.hostname, use_ssl=True, tls=tls_configuration)
+            else:
+                return Server(host=self.hostname, use_ssl=True)
+        else:
+            return Server(host=self.hostname, use_ssl=False)
+
+    def get_roles(self, groups):
+
+        groups = [group.lower() for group in groups]
+
+        return [role for role, mapping in self.role_mappings.items() if any(group.lower() in [mapped_group.lower() for mapped_group in mapping] for group in groups)]
+
+    def authenticate(self, auth_package):
+        """Authenticate a user.
+        :param auth_package: data needed to authenticate with this provider
+        :type auth_package: ADAuthPackage
+        :return: auth data
+        :rtype: AuthData
+        """
+
+        if any(c in auth_package.username for c in INVALID_SAMACCOUNTNAME_CHARS):
+            raise InvalidCredentialsException("Submitted samaccountname contained invalid characters")
+
+        conn = Connection(server=self.get_server(),
+                          user=self.domain + "\\" + auth_package.username,
+                          password=auth_package.password,
+                          authentication=NTLM)
+
+        # Connects with user; True if Valid Creds and Server Reachable
+        if conn.bind():
+            if conn.search(
+                    search_base=self.base_dn,
+                    search_filter=f'(sAMAccountName={auth_package.username})',
+                    attributes="memberOf"):
+                user_data = json.loads(conn.response_to_json())
+                user_groups = user_data['entries'][0]['attributes']['memberOf']
+
+                roles = self.get_roles(user_groups)
+                conn.unbind()
```

Comment:
```suggestion
                conn.unbind()
                
                user_groups = user_data['entries'][0]['attributes']['memberOf']
                roles = self.get_roles(user_groups)
                
```

Suggest we go ahead and unbind right away; no point in keeping open and possibly leaving it orphaned if we have run into errors w/ post-processing.